src/com/xilinx/rapidwright/design/tools/RegroupInstances.java

```diff
@@ -0,0 +1,145 @@
+                EDIFCell newCell = new EDIFCell(work, proposedInstName + EDIFTools.getUniqueSuffix());
+                EDIFHierCellInst grandParent = netlist.getHierCellInstFromName(idx == -1 ? "" : newParentName.substring(0, idx));
+                EDIFCellInst newParentInst = grandParent.getCellType().createChildCellInst(proposedInstName, newCell);
+                newParent = grandParent.getChild(newParentInst);
+            }
+            ECOTools.refactorCell(design, instance, newParent);
+        }
+    }
+
+    public static Map<String, String> parseRegroupMappingsFile(String fileName) {
+        Map<String, String> mappings = new LinkedHashMap<>();
+        int lineNum = 0;
+        for (String line : FileTools.getLinesFromTextFile(fileName)) {
+            lineNum++;
+            String tmp = line.trim();
+            if (line.trim().length() == 0 || tmp.startsWith("#")) {
+                continue;
+            }
+            String[] parts = line.split("\\s+");
+            if (parts.length > 2) {
+                throw new RuntimeException(
+                        "ERROR: Unrecognized tokens on line " + lineNum + "'" + line + "'");
+            }
+            if (parts.length == 1) {
+                // Regroup to top level
+                mappings.put(parts[0], "");
+            } else {
+                mappings.put(parts[0], parts[1]);
+            }
+        }
+        return mappings;
+    }
+
+    public static void main(String[] args) {
+        if (args.length != 3) {
+            System.out.println("USAGE: <input_design{.edf|.dcp}> <output_design{.edf|.dcp}> <regroup_mappings.txt>");
+            System.out.println("    *** Regroup Mappings File Format ***");
+            System.out.println("    <current hierarchical instance name to regroup> <new parent instance name>");
+            System.out.println("    ...");
+            System.out.println("    # For example, the following file: ");
+            System.out.println("    #   base_mb_i/microblaze_0/U0/MicroBlaze_Core_I/Performance.Core partition2");
+            System.out.println("    #   base_mb_i/microblaze_0_local_memory/dlmb_v10 partition2/microblaze_0_local_memory_2");
+            System.out.println("    #   base_mb_i/microblaze_0_local_memory/ilmb_v10 partition2/microblaze_0_local_memory_2");
+            System.out.println("    # Would result in regrouped instances:");
+            System.out.println("    #   partition2/Performance.Core");
+            System.out.println("    #   partition2/microblaze_0_local_memory_2/dlmb_v10");
+            System.out.println("    #   partition2/microblaze_0_local_memory_2/ilmb_v10");
+            return;
+        }
+
+        CodePerfTracker t = new CodePerfTracker("Regroup Instances");
+        Design design = null;
+        t.start("Read Design");
+        if (args[0].endsWith(".dcp")) {
+            design = Design.readCheckpoint(args[0], CodePerfTracker.SILENT);
+            design.unplaceDesign();
```

Comment:
Can we document and issue a warning here that using this class on placed DCPs will result in everything (including instances that are not regrouped) from being unplaced (and unrouted)?

Reply:
Am I right that unplacing is necessary because `ECOTools.refactorCell()` does not currently work for non-leaf cells which are placed? Let's leave that as a note.

Reply:
> Can we document and issue a warning here that using this class on placed DCPs will result in everything (including instances that are not regrouped) from being unplaced (and unrouted)?

Warning added.

> Am I right that unplacing is necessary because `ECOTools.refactorCell()` does not currently work for non-leaf cells which are placed? Let's leave that as a note.

Yes, I've added a note.